### PR TITLE
dut monitor: Support local port forwarding option

### PR DESCRIPTION
**Pull request by mhiramat**
Add the options for local port forwarding for each DUT on `cro3 dut monitor`. This allows user to make DUT working as gateway. e.g.

cro3 dut monitor mydut,192.168.111.222:3456

This will map some local port to 192.168.111.222:3456 via mydut.

### CODE REVIEW

File: src/dut.rs [hunk 1]
```diff
@@ -649,18 +710,26 @@ impl SshInfo {
...
     // The future will be resolved once the first connection attempt is succeeded.
-    pub async fn start_ssh_forwarding(&self, port: u16) -> Result<async_process::Child> {
-        self.start_ssh_forwarding_in_range(Range {
-            start: port,
-            end: port + 1,
-        })
+    pub async fn start_ssh_forwarding(
+        &self,
+        port: u16,
+        additional_ssh_args: Option<&Vec<&str>>,
```
Comment (brandonpollack23) [inline src/dut.rs:716]: I think option isnt necessary since it can just be empty

File: src/dut.rs [hunk 2]
```diff
@@ -649,18 +710,26 @@ impl SshInfo {
...
+            additional_ssh_args,
+        )
         .await
         .map(|e| e.0)
     }
     // Start SSH port forwarding in a given range without timeout.
     async fn start_ssh_forwarding_in_range(
         &self,
         port_range: Range<u16>,
+        additional_ssh_args: Option<&Vec<&str>>,
```
Comment (brandonpollack23) [inline src/dut.rs:732]: ditto

File: src/dut.rs [hunk 3]
```diff
@@ -627,18 +681,25 @@ impl SshInfo {
         port: u16,
         dut_port: u16,
         command: &str,
+        additional_ssh_args: Option<&Vec<&str>>,
```
Comment (brandonpollack23) [inline src/dut.rs:684]: ditto

File: src/dut.rs [hunk 4]
```diff
@@ -627,18 +681,25 @@ impl SshInfo {
...
+            "-L",
+            &format!("{}:127.0.0.1:{}", port, dut_port),
+            "-o",
+            "ExitOnForwardFailure yes",
+            "-o",
+            "ServerAliveInterval=5",
+            "-o",
+            "ServerAliveCountMax=1",
+        ];
+        let mut args: Vec<&str> = fixed_args.to_vec();
```
Comment (brandonpollack23) [inline src/dut.rs:696]: dont use an array then to_vec just use the vec![] macro

File: src/dut.rs [hunk 5]
```diff
@@ -627,18 +681,25 @@ impl SshInfo {
...
+        if let Some(opt_args) = additional_ssh_args {
```
Comment (brandonpollack23) [inline src/dut.rs:697]: remove if let once it is just a vec

File: src/dut.rs [hunk 6]
```diff
@@ -123,12 +169,20 @@ impl MonitoredDut {
...
                     self.reconnecting = false;
-                    Ok(format!(
+                    let mut status = format!(
                         "{:<31}\t127.0.0.1:{:<5}\t{}",
                         &self.dut,
                         self.port,
                         &self.ssh.host_and_port()
-                    ))
+                    );
+                    self.fwports.iter().for_each(|fwp: &PortForwarding| {
```
Comment (brandonpollack23) [inline src/dut.rs:178]: I'm not really a fan of this for_each stuff when you could just map the whole vec, produce a string for each element, then join the string using vec/iter's methods

File: src/dut.rs [hunk 7]
```diff
@@ -123,12 +169,20 @@ impl MonitoredDut {
             match child.try_status()? {
                 None => {
                     self.reconnecting = false;
-                    Ok(format!(
+                    let mut status = format!(
```
Comment (brandonpollack23) [inline src/dut.rs:172]: the below would also get read of the need for mutation

File: src/cmd/dut.rs [hunk 8]
```diff
@@ -212,9 +213,23 @@ fn run_dut_monitor(args: &ArgsDutMonitor) -> Result<()> {
...
+        let ports: Vec<&str> = raw_dut.split(',').collect();
+        let mut fwports: Vec<PortForwarding> = vec![];
+        let dut_port = port;
+
         port += 1;
+
+        if ports.len() > 1 {
+            for fwport in ports.iter().skip(1) {
+                let parts: Vec<&str> = fwport.split(':').collect();
+                let port_no: u16 = parts[1].parse().expect("Invalid port number.");
```
Comment (brandonpollack23) [inline src/cmd/dut.rs:226]: can you return the error with the string that coulddnt be parsed so whenever it is thrown that informatoin is included?

File: src/cmd/dut.rs [hunk 9]
```diff
@@ -212,9 +213,23 @@ fn run_dut_monitor(args: &ArgsDutMonitor) -> Result<()> {
     let mut targets: Vec<MonitoredDut> = Vec::new();
     let mut port = 4022;
 
-    for dut in &args.duts {
-        targets.push(MonitoredDut::new(dut, port)?);
+    for raw_dut in &args.duts {
+        let ports: Vec<&str> = raw_dut.split(',').collect();
```
Comment (brandonpollack23) [inline src/cmd/dut.rs:217]: instead of a direct string cant our command line parser handle this?

**Review by brandonpollack23 — CHANGES_REQUESTED** [review]
initial comments for now

File: src/cmd/dut.rs (hunk 9 above)
Comment (mhiramat) [reply src/cmd/dut.rs:217]: Hmm, what syntax do you expect for this feature?
Command line parser expects only `--option option_arg` or  positional arguments which separated by spaces. So I think it is not possible to specify *which* ip&port is forwarded from which DUT.

Comment (brandonpollack23) [reply src/cmd/dut.rs:217]: i guess argh doesnt support it

File: src/dut.rs (hunk 6 above)
Comment (mhiramat) [reply src/dut.rs:178]: Hm, OK. I just feel that loops on the list twice, but it maybe C programmer's thought.

File: src/dut.rs (hunk 4 above)
Comment (mhiramat) [reply src/dut.rs:696]: I've tried to use `vec![]` but compiler complained it has short lived object. Maybe I need to hold `&format!("{}:127.0.0.1:{}", port, dut_port)` somewhere.

File: src/dut.rs (hunk 1 above)
Comment (mhiramat) [reply src/dut.rs:716]: Hm, so usually Vec does not need Option, right?

File: src/dut.rs (hunk 7 above)
Comment (mhiramat) [reply src/dut.rs:172]: Got it.

File: src/cmd/dut.rs (hunk 8 above)
Comment (mhiramat) [reply src/cmd/dut.rs]: Done.

**Review by HidenoriKobayashi — APPROVED** [review]
LGTM

**Pull request merged**